### PR TITLE
Add doxygen-check pre-commit hook

### PR DESCRIPTION
**Pull request by karthikeyann**
- Adds doxygen-check pre-commit hook (verbose)
  - doxygen-check passes with warning if doxygen is not present
  - doxygen-check passes with warning if supported doxygen version 1.8.20 to 1.9.1 is not present
  - doxygen-check fails if any warning or error is thrown by doxygen
  - since pre-commit runs as part of style check, it's removed from _style.sh_
- updated Doxyfile configuration to work with versions 1.8.20 to 1.9.1
- added doxygen=1.8.20 to conda dev environment
- mentioned doxygen.sh in CONTRIBUTING.md guide
Related to https://github.com/rapidsai/cudf/issues/9373 (Add Doxygen CI check)

### CODE REVIEW

File: conda/environments/cudf_dev_cuda11.5.yml
Comment (bdice) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: I installed 1.9.2 and got these warnings.

```
warning: Tag 'OUTPUT_TEXT_DIRECTION' at line 102 of file '-' has become obsolete.             
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'COLS_IN_ALPHA_INDEX' at line 1089 of file '-' has become obsolete.              
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'LATEX_SOURCE_CODE' at line 1836 of file '-' has become obsolete.                
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'RTF_SOURCE_CODE' at line 1926 of file '-' has become obsolete.                  
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'DOCBOOK_PROGRAMLISTING' at line 2031 of file '-' has become obsolete.           
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
```

Since this is a `language: system` package that is not managed by pre-commit, we will need to make sure it works for slightly-older and current versions. At the time of writing:
- [Ubuntu 18.04 ships doxygen 1.8.13](https://packages.ubuntu.com/bionic/doxygen).
- [Ubuntu 20.04 ships doxygen 1.8.17](https://packages.ubuntu.com/focal/doxygen).
- [Ubuntu 22.04 ships doxygen 1.9.1](https://packages.ubuntu.com/jammy/doxygen).

I'd say that supporting 1.8.13 and newer is an ideal goal.

File: .pre-commit-config.yaml
Comment (bdice) [inline .pre-commit-config.yaml:110]: We need `pass_filesnames: false` to ensure this is called once, rather than once per file.
```suggestion
                language: system
                pass_filenames: false
```

**Review by bdice — CHANGES_REQUESTED** [review]
Awesome @karthikeyann. This is going to work great with a few alterations. Comments below.

File: conda/environments/cudf_dev_cuda11.5.yml
Comment (karthikeyann) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: I am using doxygen `1.8.20` because that's the version use to generate https://docs.rapids.ai/api/libcudf/nightly/

Comment (bdice) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: pre-commit is not managing this dependency and uses a system (or conda) installation of `doxygen`, whatever is on the user's path. Users of Ubuntu 22.04 who do not use a conda environment to build would likely be fetching the system-installed apt package for doxygen 1.9.1. Those users would get warnings and the pre-commit check would fail. _Regardless_ of whether users are in a conda environment, they can make use of all other pre-commit hooks because they are _self-managed_ by pre-commit itself. Because this hook cannot be self-managed by pre-commit, we need to take care of two small things:
- Users without doxygen installed must be able to make pre-commit pass locally. In this case, CI (which has doxygen installed) will catch any issues.
- Users with newer versions of doxygen must be able to make pre-commit pass locally (because we cannot pin a dependency that is not managed by pre-commit). I think we can easily adapt our configuration to be compatible with both 1.8.x and 1.9.x, and that should solve 95% of the potential problems here.

If you'd like to see a concrete solution, I can push a small commit to help resolve both issues.

Comment (bdice) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: Forgot to add -- if we solve the two problems above (no doxygen / new doxygen), it's totally fine to pin this the same way as our official documentation builds.

Comment (bdice) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: I tried running `doxygen -u` to update with 1.9.2. It works, but it leaves some options that are unsupported by 1.8.x.
```
warning: ignoring unsupported tag 'RESOLVE_UNNAMED_PARAMS' at line 527, file -
warning: ignoring unsupported tag 'SHOW_HEADERFILE' at line 600, file -
warning: ignoring unsupported tag 'WARN_IF_INCOMPLETE_DOC' at line 818, file -
warning: ignoring unsupported tag 'FULL_SIDEBAR' at line 1536, file -
warning: ignoring unsupported tag 'MATHJAX_VERSION' at line 1619, file -
warning: ignoring unsupported tag 'DOT_UML_DETAILS' at line 2398, file -
warning: ignoring unsupported tag 'DOT_WRAP_THRESHOLD' at line 2407, file -
```

We'll probably need a slightly more complicated `sed` command to filter either 1.9.x warnings or 1.8.x warnings, depending on what version we pin to in the conda env.

File: ci/checks/style.sh
Comment (karthikeyann) [reply ci/checks/style.sh:58]: pre-commit will not run doxygen if there are no file changes in `cpp/include` in that PR. It may not run always.

Comment (bdice) [reply ci/checks/style.sh:58]: The CI command above is invoked with `--all-files` for this reason. It will always run, as we intend.

Try `pre-commit run --hook-stage manual --all-files` locally and you can see for yourself.

Comment (karthikeyann) [reply ci/checks/style.sh]: removed doxygen check in style.sh, updated pre-commit to run in manual mode too.

File: conda/environments/cudf_dev_cuda11.5.yml
Comment (karthikeyann) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: used `language: conda` to install doxygen=1.8.20 and use it for pre-commit hook.
I hope, this resolved the dependency issue completely.

File: ci/checks/style.sh
Comment (karthikeyann) [inline ci/checks/style.sh:54]: Just noticed that, `HEADER_META_RETVAL` is not used in `RETVALS` list for determining the style check failure. 
Is this an oversight or intentional?

File: conda/environments/cudf_dev_cuda11.5.yml
Comment (bdice) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: Up to now, I have avoided using `language: conda` in pre-commit hooks because it requires a system-installed conda. Developers can currently run pre-commit with the system Python, without using a conda environment. It is my understanding that this is a common developer use case that we should support. I would rather try to use whatever doxygen is already on the user’s path, with fallbacks / workarounds in the script as mentioned above.

File: ci/checks/style.sh
Comment (bdice) [reply ci/checks/style.sh:54]: I think this is a bug. @vyasr Can you confirm if that was an unintentional change in https://github.com/rapidsai/cudf/commit/1ab315a1f019d1223c19b9b89ee116e0024aa254?

File: conda/environments/cudf_dev_cuda11.5.yml
Comment (karthikeyann) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: I got the requirement about system python. That's the reason for using wheel version of clang-format too!

Comment (bdice) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: Exactly! We prefer pre-commit hooks that are Python (or pip-installable binaries), and `language: system` is a fallback option that may require additional work. From the list of [Supported Languages](https://pre-commit.com/#supported-languages), we really only want `python` or `system` (or possibly `script`?). All the others require the system to have other tools installed.

Comment (karthikeyann) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: Apart from `unsupported tag` warnings, there are many bogus warning in older versions `<1.8.20`. 
Adding regex to ignore all those warnings is a hacky approach, and not recommended. I will enforce the doxygen version check in `ci/check/doxygen.sh`.

Regarding updating the Doxyfile configuration for different version, unsupported tags alone can be removed to not throw  `unsupported tag` warnings on version from 1.8.20 to 1.9.1.
For doxygen versions `>1.9.1`  there are bogus warnings again. So, we cannot support them yet. (for example, default constructor without argument is throwing warning that their parameters are not all documented)

Comment (bdice) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: > I will enforce the doxygen version check in ci/check/doxygen.sh.

That sounds good to me!

Comment (karthikeyann) [reply conda/environments/cudf_dev_cuda11.5.yml:46]: PR updated. Please re-review

File: ci/checks/style.sh
Comment (vyasr) [reply ci/checks/style.sh:54]: Yes, this is definitely a bug on my part.

**Review by bdice — COMMENTED** [review]
This still needs to exit gracefully if doxygen versions don't match (or doxygen isn't found).

File: ci/checks/doxygen.sh
Comment (karthikeyann) [reply ci/checks/doxygen.sh:15]: >What happens if doxygen is not installed?
Fails
> What happens if an incompatible version is installed?
Fails

If doxygen-check passes when doxygen is not installed or with unsupported version, I am concerned that users might be confused that local pre-commit doxygen-check passes, but pre-commit in CI failed. 
Since pre-commit swallows STDOUT and STDERR when exit code is zero, users may not know the reason for local doxygen-check pass. IMO, it's showing different behavior in CI vs local machine silently, which ideally should not happen.

**Review by bdice — APPROVED** [review]
LGTM ~once we have consensus on the exit code for local runs of pre-commit when doxygen is not installed / wrong versions.~ edit: resolved by `verbose: true` and `exit 0`. I really appreciate your hard work on this, it’s tricky to consider all the cases but it will be very nice for maintaining the quality of the docs.

**Comment by PointKernel** [comment]
Do we need to add a section in [contributing guide](https://github.com/rapidsai/cudf/blob/branch-22.08/CONTRIBUTING.md ) mentioning how Doxygen check works in cudf?

File: ci/checks/doxygen.sh
Comment (bdice) [reply ci/checks/doxygen.sh:15]: Note that it is not possible to satisfy this version requirement (1.8.20 <= version <= 1.9.1) using `apt` on Ubuntu 18.04 or 20.04. Only Ubuntu 22.04 has a compatible version (doxygen 1.9.1) and that could be upgraded to an incompatible version in the future at any time. The lack of graceful exit means that users effectively _must_ use conda to satisfy the doxygen requirement (or build doxygen from source), which we specifically sought to avoid in the past. Users should be able to make pre-commit pass with software from only a system package manager, with no conda required. Otherwise users will simply not utilize pre-commit and benefit from the majority of features that _do_ work on all systems.

Comment (PointKernel) [reply ci/checks/doxygen.sh:15]: > We already lean towards letting things pass locally and fail in CI rather than putting requirements on users to install all system dependencies

Can you please provide more details on how this decision was made? I personally find it's impractical that sometimes my local pre-commit checks passed but then the CI failed. e.g. some of the cmake format failures, if not found by the local pre-commit hook, were obvious but others took me a while to "guess" where went wrong. If users choose to contribute to the code, I think it's fair to require them to install all the proper dependencies.

> Python users in particular shouldn’t be required to install doxygen if they only make a small C++ tweak.

If they commit the "tweak", enforcing doxygen check is a reasonable requirement.

Comment (bdice) [reply ci/checks/doxygen.sh:15]: > If they commit the "tweak", enforcing doxygen check is a reasonable requirement.

My point is that the user may not touch something that would require documentation updates, but doxygen 1.8.20 would still be required to run other checks like clang-format on commit. (Unless we rely on doxygen being last in the pre-commit configuration and tell users to skip pre-commit when running for a second time. That’s annoying.)

Comment (bdice) [reply ci/checks/doxygen.sh:15]: (FWIW, this entire discussion becomes moot if someone packages doxygen in the same way that clang-format was packaged as a wheel. Then it could be managed by pre-commit rather than used as a system package. This is why we didn’t have clang-format in pre-commit until recently — it couldn’t be managed automatically for users.)

Comment (PointKernel) [reply ci/checks/doxygen.sh:15]: Right, I totally get your point here.

There are actually pros and cons in either way (exit gracefully or not). By all means, if a PR fails doxygen check, it's still the contributor's responsibility to fix the issue so the person needs to install the proper version of doxygen anyway. 

How about exiting cleanly and adding a detailed message when doxygen check fails? e.g. providing instrucions about the proper version of doxygen, how to reproduce the doxygen error locally, etc. In this way, users won't be surprised by the unexpected CI failure and know the right way to fix it.

Comment (PointKernel) [reply ci/checks/doxygen.sh:15]: > (FWIW, this entire discussion becomes moot if someone packages doxygen in the same way that clang-format was packaged as a wheel. Then it could be managed by pre-commit rather than used as a system package. This is why we didn’t have clang-format in pre-commit until recently — it couldn’t be managed automatically for users until the wheel was released.)

I actually check the clang-format wheel repo to see how much effort is needed to maintain such a project for doxygen. It requires some work to set it up, but once it's working, the maintenance seems effortless. So if it's really useful to us, we can build the "wheel" by ourselves.

Comment (bdice) [reply ci/checks/doxygen.sh:15]: > it's still the contributor's responsibility to fix the issue so the person needs to install the proper version of doxygen anyway.

I would argue that while it is the contributor's responsibility, that doesn't imply that the contributor needs to install doxygen locally. I frequently fix problems by reading the output of CI rather than reproducing the problem locally. Failing at the "local commit" stage just means a user has more dependency installation hoops to jump through before they can commit and push to see the failing output in CI. The CI logs would direct them on what to fix locally.

Comment (bdice) [reply ci/checks/doxygen.sh:15]: @karthikeyann @PointKernel I accept that there are pros and cons to any decision here. I approve this change but I expect pushback from developers (including myself in some cases) who do not use a conda environment and will see local failures due to introducing the first hard dependency that cannot be managed by pre-commit itself. 😕

Comment (vyasr) [reply ci/checks/doxygen.sh:15]: @bdice is correct, running the cmakelang hooks require that libcudf be configured because configuration downloads rapids-cmake into the build directory and rapids-cmake contains a format file that we need. That is the primary reason for that hook being designed to allow "soft" failures, and it's necessary because CMake allows users to place the build directory wherever they want. In theory we could _require_ users to set environment variables to configure the search, but we decided that was overkill given that the number of users who make CMake modifications beyond just adding a single new file to a CMakeLists.txt is extremely limited. With the current approach only users who actually make substantial CMake modifications need to care about this.

I agree with @PointKernel that it is reasonable to enforce documentation style on any C++ contribution, no matter how small, but I'm with Bradley that this doesn't imply a need for _local_ failures, as long as CI fails and the user is made aware of why. However, I think the real problem that we run into here is this one:

> I personally find it's impractical that sometimes my local pre-commit checks passed but then the CI failed. e.g. some of the cmake format failures, if not found by the local pre-commit hook, were obvious but others took me a while to "guess" where went wrong.

Ideally we want to provide users sufficient diagnostic information as to why pre-commit passed locally but then failed on CI. I think that the appropriate solution here is to change any hooks that we want to have this "silent pass" behavior to always be verbose (we can set `verbose: true` in the `.pre-commit-config.yaml` file for those hooks). It'll add a little bit of noise to the pre-commit run even when it passes (IIRC it'll print the hook id and how long the hook took to run), but I think that's an acceptable tradeoff for providing useful diagnostics. Currently when run verbosely the cmake-format and cmake-lint hooks will print which format file they are using, or if none was found. Always presenting that information to the user would help alleviate concerns over different behaviors locally and in CI. Similarly, including diagnostic information about doxygen versions in the doxygen hook (the script already has them, we just need to stop pre-commit from swallowing them) would indicate to users why they run into problems. This could be especially valuable for users that run manually (not through pre-commit) but are running a different local version than our required version. Running the pre-commit hook would immediately tell them why their local linter was producing output that the CI linter considers invalid.

Comment (bdice) [reply ci/checks/doxygen.sh]: Brilliant idea for `verbose: true` @vyasr! That's a great solution and I like the implementation @karthikeyann. Nice teamwork!

**Comment by karthikeyann** [comment]
> Do we need to add a section in [contributing guide](https://github.com/rapidsai/cudf/blob/branch-22.08/CONTRIBUTING.md?rgh-link-date=2022-06-16T13%3A08%3A31Z) mentioning how Doxygen check works in cudf?

Added a line about `ci/checks/doxygen.sh` as linter.

File: ci/checks/doxygen.sh
Comment (karthikeyann) [reply ci/checks/doxygen.sh]: Done 👍 
Thank you @bdice @PointKernel @vyasr ! I really appreciate this great discussion.

**Review by PointKernel — APPROVED** [review]
Looks great to me! Thanks @karthikeyann !

**Comment by jrhemstad** [comment]
Can we update the PR description with a little more detail on what this will actually do?

**Comment by davidwendt** [comment]
Is the required/recommended doxygen version documented anywhere?

File: ci/checks/doxygen.sh
Comment (ajschmidt8) [inline ci/checks/doxygen.sh:17]:
```suggestion
DOXYGEN_VERSION=$(doxygen --version)
```

For consistency with the rest of the script, we should switch this to use `$()` instead of ``` `` ```.

**Review by ajschmidt8 — APPROVED** [review]
Approving with one minor style suggestion.

**Comment by vyasr** [comment]
> Is the required/recommended doxygen version documented anywhere?

For that matter, is the required clang-format version documented anywhere (I do not consider environment or pre-commit files to be documentation)? We probably want both.

Also as discussed above, we should change the cmakelang hooks to also use `verbose: true`. That's not really in scope for this PR, so I'd be happy to see that done after this one is done, just don't want to see it fall off the radar.

**Comment by karthikeyann** [comment]
@gpucibot merge

File: ci/checks/doxygen.sh
Comment (jrhemstad) [reply ci/checks/doxygen.sh]: I'm not sure about how this decision played out.

To make sure I understand... 
- I want to locally reproduce a Doxygen warning failure in CI 
- I don't have the same Doxygen version installed 
- Therefore, I need to go download and install the same Doxygen version as used by CI. 

I think any solution that requires a contributor to have to go out and install some new package just to merge my PR to is a barrier to entry.

Comment (jrhemstad) [reply ci/checks/doxygen.sh]: What if we just built the right version of Doxygen from source as a dependency of the `make docs` cmake target?

Comment (bdice) [reply ci/checks/doxygen.sh]: It was my understanding that the CI failure would contain enough information about the problem and solution that a local reproduction is not necessary. Is that true?

Comment (bdice) [reply ci/checks/doxygen.sh]: Until recently, it was required to have a specific version of clang-format to locally reproduce a CI failure. This was simplified by using a pre-commit hook that can manage its own installation of clang-format. The doxygen requirement is no more stringent than what we previously required for clang-format.

Comment (jrhemstad) [reply ci/checks/doxygen.sh]: > It was my understanding that the CI failure would contain enough information about the problem and solution that a local reproduction is not necessary. Is that true?

I think it's impractical to expect this to be true. Imagine CI generated 10 warnings and I wanted to go through and iteratively address them. Do I need to submit to GitHub and wait for CI all 10 times to ensure each fix is working as expected?

> The doxygen requirement is no more stringent than what we previously required for clang-format.

Right, and I wasn't a fan then either. 

If we're going to require a specific version of a thing, we should automate getting it. We did this with `clang-format`. I think we can do this w/ Doxygen by just building it from source (depending on how onerous building it from source is).

Comment (bdice) [reply ci/checks/doxygen.sh]: I’d support exploring the build-from-source option within pre-commit. I wonder if pre-commit hooks can execute code on installation (or build on first run?). That would be equally as good as pip-installable wheels. However, there is already a model for making wheels for non-Python packages in the clang-format wheels so I might lean towards trying that way first.

Comment (bdice) [reply ci/checks/doxygen.sh]: > Imagine CI generated 10 warnings and I wanted to go through and iteratively address them. Do I need to submit to GitHub and wait for CI all 10 times to ensure each fix is working as expected?

I think all the warnings show at once. I don’t think it exits early at the first failure, so a single CI run should be enough to fix all problems. Luckily, the style check runs quickly so iteration is fast between pushing local changes and watching for CI failures.

Comment (karthikeyann) [reply ci/checks/doxygen.sh]: All the errors come at once. No early exit.

Comment (karthikeyann) [reply ci/checks/doxygen.sh]: > I’d support exploring the build-from-source option within pre-commit. 

Do you mean, build-doxygen-from-source? 
It brings doxygen's compilation problems into pre-commit installation. I would support doxygen binaries through pip-installable wheels. https://sourceforge.net/projects/doxygen/files/ is actively maintained. so it's possible to implement similar to clang-format wheels.

Comment (vyasr) [reply ci/checks/doxygen.sh]: I'm confused here.
> If we're going to require a specific version of a thing, we should automate getting it. We did this with clang-format.

By "we did it with clang-format", do you mean in the form of our conda environment? If so, that's still the case. Someone creating a custom environment for building cudf (or using compose) will get the correct version. If someone tries to use their own environment, though, they're on their own. That is also identical with how clang-format worked before.

> I wonder if pre-commit hooks can execute code on installation (or build on first run?). That would be equally as good as pip-installable wheels.

I don't know how much control you have over the installation step, but you can execute basically arbitrary code on first run to do the installation if it doesn't already exist (in some prespecified directory). I wouldn't say that it's equally as good, though. Compiling code within a pre-commit hook immediately opens you up to any sort of build failure, which is also essentially out of your control to debug.

> However, there is already a model for making wheels for non-Python packages in the clang-format wheels so I might lean towards trying that way first.

This is the better approach if we feel strongly about doxygen needing to always be available in this fashion. If we're already going to do the work to build it in a pre-commit hook, we're most of the way there to building the wheels.

Comment (karthikeyann) [reply ci/checks/doxygen.sh]: >"We did this with clang-format"

clang-format binary is packaged as pip installable repo. [ssciwr/clang-format-wheel](https://github.com/ssciwr/clang-format-wheel)). So it avoided using creating conda environment. 
My first idea is to use conda for automate installing specific version of doxygen. but it creates additional dependency to have `conda` installed in the system, which we don't like to put these additional conditions (`conda`) on the developer. 

@jrhemstad is suggesting to create a pip installable doxygen repo which can install doxygen binary, similar to clang-format.
@jrhemstad are you suggesting to build doxygen-from-source in pre-commit installation OR in pip installable doxygen repo releases?

Comment (bdice) [reply ci/checks/doxygen.sh]: I’d echo all of what @vyasr said here: https://github.com/rapidsai/cudf/pull/11076#discussion_r921514845

Building from source on the first run of the pre-commit hook is not ideal — I only proposed exploring it because it was a different approach than previous ideas. Packaging doxygen binary wheels is the way I’d like to go if we decide to move forward here.

**Pull request merged**
